sklearn/preprocessing/target.py

```diff
@@ -0,0 +1,222 @@
+        elif self.transformer is not None:
+            self.transformer_ = clone(self.transformer)
+        else:
+            self.transformer_ = FunctionTransformer(
+                func=self.func, inverse_func=self.inverse_func, validate=False)
+        self.transformer_.fit(y)
+        if self.check_invertible:
+            n_subsample = min(1000, y.shape[0])
+            subsample_idx = np.random.choice(range(y.shape[0]),
+                                             size=n_subsample, replace=False)
+            diff = np.abs((y[subsample_idx] -
+                           self.transformer_.inverse_transform(
+                               self.transformer_.transform(y[subsample_idx]))))
+            if np.sum(diff) > 1e-7:
+                raise ValueError("The provided functions or transformer are"
+                                 " not strictly inverse of each other.")
+
+    def fit(self, X, y, sample_weight=None):
+        """Fit the model according to the given training data.
+
+        Parameters
+        ----------
+        X : {array-like, sparse matrix}, shape (n_samples, n_features)
+            Training vector, where n_samples is the number of samples and
+            n_features is the number of features.
+
+        y : array-like, shape (n_samples,)
+            Target vector relative to X.
+
+        sample_weight : array-like, shape (n_samples,) optional
+            Array of weights that are assigned to individual samples.
+            If not provided, then each sample is given unit weight.
+
+        Returns
+        -------
+        self : object
+            Returns self.
+        """
+        # memorize if y should be a multi-output
+        self.y_ndim_ = y.ndim
+        if y.ndim == 1:
+            y_2d = y.reshape(-1, 1)
+        else:
+            y_2d = y
+        self._validate_transformer(y_2d)
+        self.estimator_ = clone(self.estimator)
+        self.estimator_.fit(X, self.transformer_.transform(y_2d),
+                            sample_weight=sample_weight)
+        return self
+
+    def predict(self, X):
+        """Predict using the base estimator, applying inverse.
+
+        The estimator is used to predict and the ``inverse_func`` or
+        ``inverse_transform`` is applied before returning the prediction.
+
+        Parameters
+        ----------
+        X : {array-like, sparse matrix}, shape = (n_samples, n_features)
+            Samples.
+
+        Returns
+        -------
+        y_hat : array, shape = (n_samples,)
+            Predicted values.
+
+        """
+        check_is_fitted(self, "estimator_")
+        pred = self.transformer_.inverse_transform(self.estimator_.predict(X))
+        # if y is not a multi-output, it should be ravel
+        if self.y_ndim_ == 1:
+            return pred.ravel()
+        else:
+            return pred
+
+    def score(self, X, y, sample_weight=None):
+        """Returns the coefficient of determination R^2 of the prediction.
```

Comment:
Should state here that scoring is performed in the original, not the transformed, space.